### ISSUE TITLE
Provide UI to schedule queries

### ISSUE THREAD

**Issue by n-riesco** [plotly/falcon#425]
Requirements:
- list saved scheduled queries
- create a scheduled query and set:
  - query interval
  - or query time and/or date
- remove a scheduled query

**Comment by nicolaskruchten** [comment]
Basic proposed work plan as sent to Dawn Labs:

Basically we have the query UI in the Cloud/On-Prem product and we want to remove it and have it live inside Falcon. From what I can tell, Falcon currently persists information about queries to disk and handles the scheduling using a really primitive "every N" counter, and exposes that information via API under /queries but there is no way to see what queries are saved and when they will execute etc.

Consequently, I'd like to start this project (step 1) by having you guys design and implement, as a set of PRs into https://github.com/plotly/falcon-sql-client/, a view into what Falcon currently has persisted in terms of queries, based on the rough mockups in https://github.com/plotly/streambed/issues/10671

The next step (2) would be to add the schedule-creation UI in Falcon itself.

The next step (3) would be to upgrade Falcon's scheduling logic (and the corresponding UI, so this would be a backend+frontend project) to be more cron-like: ability to schedule things on given days, at given times etc.

**Comment by n-riesco** [comment]
I'll document briefly the current API for `/queries`:
- `GET /queries` returns an array of scheduled queries:
   - an scheduled query is an object with `{requestor, fid, uids, refreshInterval, query, connectionId}`
   - `requestor`, `fid` and `uids` are those defined in https://api.plot.ly/v2/
- `GET /queries/:fid` returns the scheduled query for the given `fid`
- `DELETE /queries/:fid` deletes the scheduled query for the given `fid`
- `POST /queries` with parameters `{requestor, fid, uids, refreshInterval, query, connectionId}` schedules the given query
- `POST /queries` with parameters `{filename, requestor, refreshInterval, query, connectionId}` creates grid named `filename` and schedules the given query

This API is implemented [here](https://github.com/plotly/falcon-sql-client/blob/52a78ed13b196024dc2e985ebbd1b1780e84677a/backend/routes.js#L637-L730).

**Comment by briandennis** [comment]
Thanks for the summary @n-riesco! To build on that, here is our outline of the existing query scheduling implementation. (Let me know if any of this is inaccurate)

## Existing Functionality

Currently scheduled queries are persisted with the following attributes:

- requested user
- `fid`/`uid`s for the grid
- query string
- connection id
- refresh interval

The refresh interval is currently stored as a number (in seconds) and queries are scheduled using `setInterval` at the specified duration.

And here's how we are thinking that can be adapted to support the new UI feature set:

## Proposed Changes

- store refresh intervals instead as [cron expressions](https://en.wikipedia.org/wiki/Cron#CRON_expression)
  - supports more flexible intervals (run on a given day, given time)
  - since it's a standard format, it allows us to leverage existing libraries (for example [node-schedule](https://github.com/node-schedule/node-schedule#cron-style-scheduling) and [pretty-cron](https://github.com/azza-bazoo/prettycron))
- persist the following query information to support additional frontend features from the mocks
  - row count
  - time of last and next executions
  - status (scheduled, running, or error)
- add API support for updating existing queries - changing the schedule for instance

Additionally, after exploring the mocks and thinking about different workflows, we think it would be a good idea to support these features as well:

- persist the following additional query info:
  - execution duration
  - a name/description to associate with the query
  - list of tags to associate with the query
  - error message if applicable
- add API support for on-demand triggering of a scheduled query (arbitrarily refresh the data/grid without needing to create a new query)

@n-riesco @nicolaskruchten let me know what you think about this approach! Any feedback greatly appreciated

**Comment by n-riesco** [comment]
@briandennis 

> since it's a standard format, it allows us to leverage existing libraries (for example `node-schedule` and `pretty-cron`)

`node-schedule` looks good to me

> add API support for updating existing queries - changing the schedule for instance

`POST /queries` does that (if there is a scheduled query associated with `fid`, [QueryScheduler#scheduleQuery](https://github.com/plotly/falcon-sql-client/blob/52a78ed13b196024dc2e985ebbd1b1780e84677a/backend/persistent/QueryScheduler.js#L78) deletes it, before storing the new query)

> add API support for on-demand triggering of a scheduled query (arbitrarily refresh the data/grid without needing to create a new query)

`POST /queries` can also do that, since it invokes [QueryScheduler#queryAndUpdateGrid](https://github.com/plotly/falcon-sql-client/blob/52a78ed13b196024dc2e985ebbd1b1780e84677a/backend/routes.js#L685) before storing the scheduled query.

Umm... (after some thinking) I understand this behaviour of `POST /queries` is undesired for cron queries. We don't want to trigger `QueryScheduler#queryAndUpdateGrid` when a query is created or updated (it's the cron expression what determines when `QueryScheduler#queryAndUpdateGrid` is invoked).

I'd suggest that the new API support is implemented at a different entrypoint (e.g. `/schedule`).

---

@scjody (with on-prem needs in mind) do we need to keep backwards-compatibility? Do we need to ensure that scheduled queries stored in `~/.plotly/connector/settings.yml` by older versions of Falcon can still be run by Falcon after implementing this work?

---

> @n-riesco @nicolaskruchten let me know what you think about this approach!

LGTM

**Comment by nicolaskruchten** [comment]
I love the UX direction you're headed in @briandennis with the force-run and the extra persisted data etc but it's kind of out of scope for the project I've laid out, so I'd like to wait on some of these discussions.

* If you look at the plan I laid out at the top, I was hoping we would start with as step 1 "a view into what Falcon currently has persisted in terms of queries" which should be an easy lift and require no back-end changes, so I'd like us to focus on getting that PR'ed and merged please :)
* For steps 2 & 3 the main objective is to be able to schedule queries through the UI (again without backend changes) and then to be able to schedule them with a finer grain than `setInterval`, which should in principle require no more changes to the backend than switching from an int to a cron-string, if we accept the current "overwrite and run right away when saving" behaviour
* I would like to move discussion about extra persisted data to a separate Github Issue, to be worked on once we have a UI to plug that data into, and ditto for the force-run issue. Note that right now we don't support names or tags for scheduled queries either :)

Re the [proposed UX in Figma](https://www.figma.com/proto/TvIaY2rQfdaEMMiy1bC7CW9f/falcon?node-id=83%3A540&scaling=min-zoom&redirected=1), I like it as a long-term direction but I'd like to see what the short-term would look like. The biggest UX concern I have is not so much around the metadata etc but more around clear naming/signalling of what is happening, for the user. 

Right now in the "Query" tab it's pretty straightforward: you specify and run a query, and you see the output right away inline as a table or chart and you can export it. There are basically 4 kinds of export we support today, with non-uniform UX: 

1. you can save the data or the chart to local disk via the Export subtab
2. you can save the chart to local disk via the modebar camera icon
3. you can one-time export the chart to Chart Studio via the modebar save icon OR via the Export subtab
4. you can one-time export the data to Chart Studio via the Export subtab

current Chart subtab, with modebar:

![image](https://user-images.githubusercontent.com/203523/41917416-5051832c-7928-11e8-93ab-aa975e8dbe11.png)

current Export subtab: 

![image](https://user-images.githubusercontent.com/203523/41917427-576fe8c4-7928-11e8-85c3-e2623918d5cd.png)

What we are calling "scheduled queries" are actually a lot closer to a fifth "Send CSV to Chart studio on a schedule, overwriting the dataset each time" so I might actually favour the launch-point for scheduling a query to be located in the Export subtab, with some uniform message cleanup. 

A big hole in our UX right now is that we don't support the 6th option suggested by symmetry, which would be to "send the chart to chart studio on a schedule" but we'll need to tackle that one later :)

**Comment by scjody** [comment]
> @scjody (with on-prem needs in mind) do we need to keep backwards-compatibility? Do we need to ensure that scheduled queries stored in `~/.plotly/connector/settings.yml` by older versions of Falcon can still be run by Falcon after implementing this work?

Yes, backwards compatibility is one of our [On-Prem commitments to customers](https://github.com/plotly/dev-docs/blob/master/on-prem/developer-setup.md#our-commitments-to-customers). If you need to change the format, I suggest looking for an old file when the `falcon` container starts and converting it to the new format if needed. (Then delete the old file so it doesn't get converted next time.)

If it isn't reasonable to do this, we need approval from the sales team and we need to give notice one On-Prem version in advance. In other words we'd need to inform people in the 2.5.0 release notes that things are changing for 2.6.0.

**Comment by jakedex** [comment]
I definitely agree with your UX concern, @nicolaskruchten, but I'm not sure if moving the 'schedule' button into the export context is the right solution.

IMO it makes more sense to schedule a query than to create an 'updating dataset', especially in the context of the 'schedule' tab. I think we could simply disable the 'schedule' button when the user isn't logged in and add an onhover tooltip to further explain the button's function.

On a related note, here's an updated prototype that is properly scoped to this stage of the project: https://www.figma.com/proto/TvIaY2rQfdaEMMiy1bC7CW9f/falcon?node-id=224%3A1&scaling=min-zoom

If it's alright, I think it's best to pass on bringing this up to hifi—given that we'll be using quite a few of the prexisting components/styles, it seems like a waste of time to spend the time recreating these in Figma.

**Comment by nicolaskruchten** [comment]
OK, let's skip the hifi mocks and head directly to the PRs :)

**Comment by nicolaskruchten** [comment]
But even if we still have the 'schedule' button next to 'run' I think it makes sense to have another one under 'export'.

**Comment by nicolaskruchten** [comment]
Something that just came to mind: when clicking "schedule query" how should we handle the target? Should every new scheduled query target a new dataset in the target Plotly instance? Or should the user have the choice to overwrite an existing FID? what if they delete a previously scheduled query that was sending data to a dataset upon which one or more charts depend? They would likely want to create a new scheduled query that would target the existing dataset, right?

**Comment by nicolaskruchten** [comment]
Another semi-related thought: when editing scheduled queries, how/will the user be able to preview the changes to their SQL? Will they have to build/preview a query in the Query tab and then find it in the Schedule tab and paste the SQL in? Or will they be able to find it in the Schedule tab and somehow be able to 'edit in' the Query tab? I'm thinking of simple/common situations where someone edits the query blindly and inserts a syntax error and has no easy way of finding that out until the scheduled query fails (without any push notification!).

**Comment by jakedex** [comment]
That's a good point. The more that I think about this, I agree that something like a "Dynamic Dataset" is the right approach. In the meantime, I think something like "Upload Dataset (auto-updating)" will work but this is definitely something we'll want to make more coherent in the next stage.

**Comment by jakedex** [comment]
re grids: The user should have the choice to select/create a target when scheduling a query, and I think this will make more sense once we're referring to these as dynamic datasets. 

re editing scheduled queries: At this stage, I don't think users should be able to edit the query field (as per the mocks)—they'll just have to delete and then create a new scheduled query. On that thought though, in the future, we could add a toggle (that's enabled by default) to check for errors before creating the scheduled query.

**Comment by n-riesco** [comment]
How about `Schedule Update`?

**Comment by nicolaskruchten** [comment]
Re backwards compatibility, we should think about the implications of old/new versions of the Frontend, Backend and file formats interacting together. In the short run Plotly Cloud will keep using the 'interval' method to talk to Falcon so we have to support an "old frontend/new backend" permutation. Additionally, upgrading Falcon means we have to support "old file format/new backend" permutation.

More exotic permutations like "new file format/old backend" are worth a think also :)

**Comment by n-riesco** [comment]
@nicolaskruchten 

Re/ **backwards compatibility**

We already discussed this on a slack call, but I'll write it here. My position (from a maintainer's point of view) is that we avoid breaking backwards-compatibility by extending formats rather than redefining them.

For example, a query is currently formatted as *{requestor, fid, uids, **refreshInterval**, query, connectionId}*, where `refreshInterval` is the integer passed to `setInterval`.

For this issue, we need to store an interval in `cron` format. I suggest we do it in `cronInterval`, so that we can support both formats, the old format:  *{requestor, fid, uids, **refreshInterval**, query, connectionId}* and the new one:  *{requestor, fid, uids, **cronInterval**, query, connectionId}*.

**Comment by nicolaskruchten** [comment]
Yes agreed.

But what about a new Falcon seeing a request with the old interval format?
It should accept it, right? What format does it write? Old Falcon seeing a
file written with a new Falcon? And so on

**Comment by n-riesco** [comment]
@nicolaskruchten It'll use the old format, because the new format is unable to reproduce the old behaviour.

**Comment by n-riesco** [comment]
To be clear:
- `refreshInterval` (i.e. old format) defines an interval that starts when Falcon's startup
- `cronInterval` (i.e. old format) can't express what Falcon's startup time is.

**Comment by nicolaskruchten** [comment]
OK, so basically an entry in the yaml file should only have one of `refreshInterval` or `cronInterval`, but never both, and a "new Falcon" should honour both modes for its scheduling and honour both modes when updating a query, depending on the parameters passed in by the UI: new UI sends only `cronInterval` and old UI (hosted on Plotly Cloud/Enterprise) sends only `refreshInterval`? I could live with that.

What happens if an "old Falcon" boots up and sees an entry with `cronInterval` and not `refreshInterval`? This should be testable today.

**Comment by nicolaskruchten** [comment]
PS: I'm assuming that the idea you're proposing is that both types of entries can live in the same yaml file, right? Please correct me if I'm mistaken!

**Comment by n-riesco** [comment]
@nicolaskruchten

> OK, so basically an entry in the yaml file should only have one of refreshInterval or cronInterval, but never both, and a "new Falcon" should honour both modes for its scheduling and honour both modes when updating a query, depending on the parameters passed in by the UI: new UI sends only cronInterval and old UI (hosted on Plotly Cloud/Enterprise) sends only refreshInterval?

yes, that's the idea.

> What happens if an "old Falcon" boots up and sees an entry with cronInterval and not refreshInterval? This should be testable today.

That's equivalent to `setInterval(runQuery, NaN)`. I.e. it would run the query as soon as the previous query has completed.

Why is this significant?

> I'm assuming that the idea you're proposing is that both types of entries can live in the same yaml file, right? Please correct me if I'm mistaken!

yes, that's the idea.

**Comment by nicolaskruchten** [comment]
> Why is this significant?

I just want to make sure we know what would happen if e.g. someone creates a query with an old falcon, then upgrades and saves it, then downgrades. Seems very unlikely but still, the behaviour you're describing isn't great... Could we say that if `cronInterval` is defined, then we set `refreshInterval` to some enormous number instead, to avoid this?

**Comment by nicolaskruchten** [comment]
Other than that, thanks for thinking through the permutations with me, the scheme laid out above makes sense to me. @mfix22 what do you think?

**Comment by briandennis** [comment]
Yep, this is exactly how we were thinking it would work if we need to account for the situation where "old falcon" runs into a "new falcon" query file 👍 

We can calculate a ballpark `refreshInterval` fallback based off the `cronInterval` or just use some arbitrary value depending on how important you think accuracy is here.

**Comment by briandennis** [comment]
Is it safe to assume that all existing scheduled query intervals will conform to one of the options exposed through chart studio? (i.e. every minute, 5 minutes, hourly, daily, or weekly) 

One issue with having to support both formats is that there are some intervals which are difficult to express through cron expressions. For example, a `refreshInteval` that comes out to 45 days (`refreshInteval` = 45 * 24 * 60 * 60 = 3888000) would not be able to be easily mapped to a cron expression. This is relevant for the implementation of the scheduler because most libraries support scheduling jobs via cron expressions but do not support the "every _n_ seconds" format. (@n-riesco maybe you know of one that supports both?)

Fortunately all of the chart studio options are easily mappable to cron, but if we have to account for any old `refreshInterval` it will be more difficult to utilize an off-the-shelf scheduling solution.

**Comment by n-riesco** [comment]
@scjody Please, see the comment above. Do we need to keep backwards-compatibility for any refresh interval? or only those intervals offered by Chart Studio?

**Comment by n-riesco** [comment]
@briandennis

> One issue with having to support both formats is that there are some intervals which are difficult to express through cron expressions.

Can this be done without converting `refreshInterval` into `cronInterval`?

I had imagined something  like this woudl work:
```js
if (query.hasOwnProperty('refreshInterval')) {
    setInterval(runQuery, query.refreshInterval);
} else {
    schedule.scheduleJob(query.cronInterval, runQuery);
}
```

**Comment by briandennis** [comment]
@n-riesco yes, something like that would definitely work. Just trying to avoid making it more complex than we need to. If we can unify under a single scheduler API, that would make things easier down the road if we want to track additional scheduling metadata (time of next execution for example).

**Comment by n-riesco** [comment]
@briandennis Here's a quick untested idea:

```js
let job;
if (query.hasOwnProperty('refreshInterval')) {
    scheduleRefreshInterval();

    function scheduleRefreshInterval() {
        runQuery();
        job = schedule.scheduleJob(new Date(Date.now() + query.refreshInterval), scheduleRefreshInterval);
    }

} else {
    job = schedule.scheduleJob(query.cronInterval, runQuery);
}
```

**Comment by scjody** [comment]
I'm not aware of any supported way to set a refresh interval other than through the streambed UI.

@n-riesco Can you confirm if that's the case? If so I'm OK with only supporting those intervals.

**Comment by scjody** [comment]
@n-riesco I'm happy _not_ supporting backwards compatibility in those cases.

**Comment by nicolaskruchten** [comment]
I'm pretty comfortable with closing out this issue and shuffling the outstanding stuff we discussed into the Stage 5 issue #492 ... objections?
